File: BackEnd/PHP/laravelshop/app/Http/Controllers/Api/V1/UsedMarket/UsedMarketController.php
<?php

namespace App\Http\Controllers\Api\V1\UsedMarket;

use OpenApi\Attributes as OA;

use App\Http\Controllers\Api\V1\ApiController;
use App\Modules\UsedMarket\Services\UsedMarketService;
use Illuminate\Http\Request;

#[OA\Tag(name: 'UsedMarket')]
class UsedMarketController extends ApiController
{
    public function __construct(private readonly UsedMarketService $service) {}

    public function productPrice(int $id)
    {
        return $this->success($this->service->productPrice($id));
    }

    public function categoryPrices(Request $request, int $id)
    {
        return $this->success($this->service->categoryPrices($id, (int) $request->query('page', 1), (int) $request->query('limit', 20)));
    }

    public function estimateBuild(Request $request, int $buildId)
    {
        return $this->success($this->service->estimateBuild($request->user(), $buildId));
    }
}

File: BackEnd/Kotlin/kotlin-ktor-gradle-exposeddao-postgresql/src/main/kotlin/com/pbshop/kotlin/ktor/gradle/exposeddao/postgresql/usedmarket/UsedMarketController.kt
package com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.usedmarket

import com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.common.pbActor
import com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.common.respondStub
import com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.security.PbRole
import com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.security.requireAnyRole
import io.ktor.server.application.call
import io.ktor.server.routing.Route
import io.ktor.server.routing.get
import io.ktor.server.routing.post

class UsedMarketController(
    private val service: UsedMarketService,
) {
    fun Route.register() {
        get("/used-market/products/{id}/price") {
            call.respondStub(service.productPrice(call.pathInt("id")))
        }
        get("/used-market/categories/{id}/prices") {
            call.respondStub(
                service.categoryPrices(
                    categoryId = call.pathInt("id"),
                    page = call.request.queryParameters["page"]?.toIntOrNull() ?: 1,
                    limit = call.request.queryParameters["limit"]?.toIntOrNull() ?: 20,
                )
            )
        }
        post("/used-market/pc-builds/{buildId}/estimate") {
            call.requireAnyRole(PbRole.USER)
            call.respondStub(service.estimateBuild(call.currentUserId(), call.pathInt("buildId")))
        }
    }

    private fun io.ktor.server.application.ApplicationCall.currentUserId(): Int = pbActor().userId ?: 0

    private fun io.ktor.server.application.ApplicationCall.pathInt(name: String): Int = parameters[name]?.toIntOrNull() ?: 0
}
